### ISSUE TITLE
Potential slashing-issues

### ISSUE THREAD

**Issue by opsecx** [anoma/namada#3958]
Ok so, I was looking through some data tables from the shielded expedition, and came across something that could be a potential issue when it comes to slashing:

When looking at the records of "evidences" (see cometbft docu, for instance https://docs.cometbft.com/v0.38/spec/core/data_structures) recorded by namadexer for the shielded expedition, there is a pleasantly low number of recorded evidences of infractions, meaning most validators during the testnet behaved well and avoided double-signs and such. By the way, we note that in the namada specs, the following is listed as reasons for slashing:

"	•	Equivocation in consensus
    * voting: a validator has submitted two votes that are conflicting
    * block production: a block producer has created two different blocks for the same height
* Invalidity:
    * voting: validators have voted on invalid block
    * block production: a block producer has produced invalid block (when supported by Comet)
Unavailability is not considered an offense, but a validator who hasn't voted will not receive rewards.
Once an offense has been reported:
1. Kicking out
2. Slashing
* Individual: Once someone has reported an offense it is reviewed by validators and if confirmed the offender is slashed.
* cubic slashing: escalated slashing"

(https://specs.namada.net/modules/proof-of-stake/objects-and-txs#slashing)

So with this in mind, as noted, the number of mishaps over the entirety of the shielded expedition is quite low. However, there is a concerning trend which is what this bug/case-report is about, namely that: Almost all of the infractions during Shielded Expedition, of which some validators commit multiple/many infractions, occur in clusters around certain blocks. Moreover, these blocks are all halted blocks, ie blocks around which there have been chain halts:

I will not publicize the data on which validators were involved, as I suspect it would be put to bad use, but summarized, this is the picture: 

height, validators involved, number of infractions
4, 1, 1
5, 16, 63
8672, 4, 207
42851, 5, 27
90045, 3, 33
263384, 1, 1

(the detailed data is of course available on request from heliax/anoma or other parties who can convince me they have good motives for wanting it)

Except for a couple of outliers (1-2), which probably would be caused by bad configs or similar, the clustering pattern of these mishaps around halt-blocks, by validators who are unlikely to have been working nefariously in coordination, suggest that either:

1. something with these infractions cause the halts, or
2. (more likely in my opinion) the halts (or other external events around same) cause some clients to double-sign through no fault of the operators

The implications of this is potentially chilling: 

The Namada specs lists the reasons for especially cubic slashing, which kicks in if more validators misbehave at the same time, as being to discourage collusion and sybil-like behavior. If these validators, who there is little indication of having colluded during the SE, are all punished around the same blocks (in a compounded fashion, ie being more severely punished because others did the same during same timeframe), that means there are potential issues with the logic behind this, and ultimately the cubic slashing could unfairly cause loss of tokens due to protocol or client issues. (I should note I have not found concretely whether these infractions lead to slashing, as I don't know if this data is available. I'll keep looking into this and post if I find it)

I've been discussing this with rigorous some before releasing this report (thanks!) who has provided some valuable input, namely that there may be an issue around a replay-mechanism based on some errors reported in the discord around one of these blocks. (this could be one of many possible explanations for what could have gone wrong)

It is unfortunately beyond our scope to dig out the concrete causes of this, particularly as the network in question is dead, but I wanted to get this submitted here for consideration. Wavering between submitting to the bounty-reports email (where unfortunately there is no way of replies, acknowledgement of submitted issues, nor verdicts on matters submitted), I decided to publish here instead. If there turns out to be anything in this, I would appreciate if it is still considered for a bounty. (I should also note that if the issue has been fixed, the wider problem still remains, namely that client/ledger issues can cause unfair, cubically compounded, slashing)

**Comment by Fraccaman** [comment]
I think this is an inherited mechanism of PoS systems. If there is a bug in the code, and the bug makes validators produce/validate 2 different versions of a block, as soon as the issue is fixed somebody is gonna get slashed (since they signed an invalid block). Not totally sure if this might be an answer to your issue.

**Comment by opsecx** [comment]
> I think this is an inherited mechanism of PoS systems. If there is a bug in the code, and the bug makes validators produce/validate 2 different versions of a block, as soon as the issue is fixed somebody is gonna get slashed (since they signed an invalid block). Not totally sure if this might be an answer to your issue.

I don't think this would explain how we see these slashes around four different blocks (as listed above), some of which did not have code-upgrades involved?

**Comment by brentstone** [comment]
@opsecx what do you mean by "four different blocks"? Your list above has 6.

Also @opsecx and I chatted for a bit about this the other day, and I'm not too concerned. Some observations on my end:
- the issue with a validator suffering multiple infractions in a single block was addressed in #2574, which was merged into the SE software after block `90045` and before `263384` (notice a single infraction in the latter block).
- I find the fact that there are only a small handful of validators who infracted (after the block 5 issue) in each of the other instances to be generally encouraging that there is a good chance these operators were double-signing and that there isn't some bug
- I do not think there is any evidence of an issue with a "replay-mechanism" after seeing the discord messages. In the messages, the word "replay" appeared in log lines (from CometBFT I think) when a node was trying to sync it seemed. This seems like coincidence of the word "replay" appearing.
- The software has also changed a lot since then (> 10 releases) so it would be hard to gauge if there was some bug there that is still present in the current software.

Overall, I'm not concerned, but still good to keep this observation in mind going forward and not 100% dismiss everything here yet. Thx @opsecx!

**Comment by opsecx** [comment]
>@opsecx what do you mean by "four different blocks"? Your list above has 6.

@brentstone sorry I was being convoluted there. I consider the last one a potential outlier since only a single validator is involved, and block 4 is so close to block 5 I consider that to be possibly part of the same event. Ie four main clusters of multiple validators being involved around the same block (four clusters of interest).

**Comment by opsecx** [comment]
> @opsecx what do you mean by "four different blocks"? Your list above has 6.
> 

@brentstone sorry I was being convoluted there. I consider the last one a potential outlier since only a single validator is involved, and block 4 is so close to block 5 I consider that to be possibly part of the same event. Ie four main clusters of multiple validators being involved around the same block (four clusters of interest).

> - I find the fact that there are only a small handful of validators who infracted (after the block 5 issue) in each of the other instances to be generally encouraging that there is a good chance these operators were double-signing and that there isn't some bug

I am still skeptical that so many validators (and different ones on each of these halt blocks) would have been misconfigured into double-signing. What sort of misconfigurations would one imagine could cause such a scenario? Is there any chance we could do some testing with on-purpose chain halts?


> - The software has also changed a lot since then (> 10 releases) so it would be hard to gauge if there was some bug there that is still present in the current software.

Understood.
> 
> 
> Overall, I'm not concerned, but still good to keep this observation in mind going forward and not 100% dismiss everything here yet. Thx @opsecx! 

Sure thing!

**Comment by McDaan** [comment]
> I am still skeptical that so many validators (and different ones on each of these halt blocks) would have been misconfigured into double-signing. What sort of misconfigurations would one imagine could cause such a scenario? Is there any chance we could do some testing with on-purpose chain halts?

The large number of infractions was due to double-signing on prevotes if I recalll (or precomits) - this is in thery already addresses by the PR @brentstone mentioned above. The situation was that someone from the team said during one of the halts (prob the one with most infractions) that it wasn't necessary to restore `priv_validator_state.json` file, so that confused some operators and things ended up that way.

**Comment by opsecx** [comment]
> > I am still skeptical that so many validators (and different ones on each of these halt blocks) would have been misconfigured into double-signing. What sort of misconfigurations would one imagine could cause such a scenario? Is there any chance we could do some testing with on-purpose chain halts?
> 
> 
> 
> The large number of infractions was due to double-signing on prevotes if I recall (or precommits) - this is in theory already addressed by the PR @brentstone mentioned above. The situation was that someone from the team said during one of the halts (prob the one with most infractions) that it wasn't necessary to restore `priv_validator_state.json` file, so that confused some operators and things ended up that way.
> 
> 
> 
> The positive thing is that the issue of multiple potential infractions within a same block was fixed due to that scenario happening.
> 
> 

Ok, that's useful context, thanks. At least it explains why 16 validators at once would misconfigure. Still leaves it puzzling with the other incidents, but those are a bit less remarkable.
